flybywiresim/a32nx
# feat(mcdu): qol improvement - keyboard input to mcdu scratchpad

Fixes #4862

## Summary of Changes

Click the MCDU's screen to activate keyboard input mode to the scratchpad, you may now enter letters, numbers, dots and slashes with the keyboard, rather than clicking on the buttons. Function keys also can be used to input functions (L1-L6/R1-R6)

- [x] Able to input text and numbers into the MCDU with keyboard, by clicking on the MCDU screen to put it in "focus"
- [x] Cockpit buttons are pressed with keyboard input
- [x] Visual indicator that input is being accepted by MCDU
- [x] Options for MCDU keyboard entry
- [x] Normal Sim Keyboard Events (i.e. Camera) will be disabled while MCDU is in focus.
- [x] A timer can be set to automatically lose focus from the MCDU after a set period
- ~~Arrow key functionality - Arrow keys are currently being hijacked by the options menus... thanks Coherent.~~

## Screenshots (if necessary)



[Video](https://user-images.githubusercontent.com/15316958/121357298-46ff9680-c964-11eb-999b-95cd250006cc.mp4)




## References





## Additional context

QOL improvement.


Discord username (if different from GitHub): 2Cas #1022

## Testing instructions

Known Issues:
* Arrow keys + Esc are not captured
* CLR held (>2s) - Use Shift+Backspace instead of backspace.
* Camera is locked - hence the need for default timeout of 60s in case the screen somehow is unclickable. 
* Clicking on screen + using camera keybind keys will cause MCDU to move out of the view, as the camera is locked it is not possible to click on the screen to unfocus the MCDU - Use keyboard shortcuts CTRL+Z or press ALT to unfocus. 
* Keybindings are currently optimized for ANSI/ISO-UK keyboard layouts. ISO-DE/ISO-NORD users will notice that slash keys on their keyboards may not function as expected. This is an unfortunate limitation of the Coherent Webkit based JS Engine. The >/< (Int backslash) and # (or '/*) keys  are used as substitutes instead. Note that the numpad will work as expected regardless of keyboard layout. 

1. Ensure MCDU functions normally as a baseline
2. Attempt keyboard input without clicking on the MCDU. Nothing should happen.
3. Attempt keyboard input after clicking on MCDU. Nothing should happen (still).
4. Go into MCDU Options > Options > Realism > Keyboard Input. Select ALLOW INPUT. You will have to exit to the menu and reload the flight for this change to occur (consistent with units change from kg/lb). 
5. Click on the MCDU screen, and attempt keyboard input.. The scratchpad should be highlighted, and the title should be highlighted in cyan as well. Letters, numbers, forward slashes and periods (.) should appear on the scratchpad. Backspace should act with the same behavior as the CLR button on the mcdu. **Pressing Shift+Backspace for 1 second should function the same as holding down the CLR on the mcdu for 2 seconds.**

NOTE: Focusing the MCDU for keyboard input will stop all sim keybinds from working.  Most notably, you will not be able to move the camera. Unfocusing should return control (i.e. camera)

Invalid +/- input: -+, -+-+-+, ++-+-+
Valid +/-  input: +-, +++++-, +++-, - - - - - - -

6. Attempt to flood the MCDU and break it with keyboard input somehow. Slam the keyboard if you have to. Check that there are no inputs that would be impossible using the manual method (clicking on buttons). 
7. Check that the normal input method still works (using cockpit mcdu buttons to input into scratchpad)
8. Check that each button (on both MCDUs) moves and reacts to the keyboard. 
9. Check that all the pages are still functioning as intended. Check for visual UI glitches.
10. Check that inputting data from the scratchpad still  works as you would expect.
11. Double check that no inappropriate inputs that would not be possible with the usual method (clicking on the buttons) are possible.
12. Clicking again on the MCDU should disable keyboard input, **or press ALT or CTRL+Z**. The scratchpad should no longer be highlighted. Note: By default, there is a 60 second timeout on keyboard input. If no inputs are received in this time, the scratchpad will unfocus itself. This is intentional behaviour. 
13. Now test the aforementioned timeout function. Make an input, then wait for a minute. The scratchpad should unfocus itself. Check that keyboard input is no longer possible.
~~14. Go back to MCDU options > Options > Realism > MCDU Timeout (Seconds) and set the timeout to 0s. NONE* should appear~~
~~15. Now the scratchpad should never unfocus. Ensure that this is the case by waiting past the default timeout length. This should happen if both keyboard input is entered, or not.~~
16. Set the timeout to a custom duration (i.e. 5 seconds). Ensure that the scratchpad will lose focus after this time. This should happen if both keyboard input is entered, or not. The valid range should be 5-120 seconds.
17. Go into the options. Disable MCDU keyboard input (NONE). Reset the MCDU if prompted. Check that previous behavior is retained.

Please note these testing instructions are just here to *inform* your testing procedure, don't just stick to this. Do try your best to break things.  


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
